### DISCUSSION TITLE
Is junction tree algorithm available? Why does BeliefPropagation take too long?

### DISCUSSION THREAD

**Discussion by ferencbartok** [pgmpy/pgmpy#1875] [Q&A] [answered]
Hi,

I'm writing this as sort of a follow-up to the questions in this discussion: https://github.com/pgmpy/pgmpy/discussions/1746 .
I'm experimenting with pgmpy again and I'd have a few more questions.

Is Junction tree algorithm available in pgmpy? I could not find it in the docs, just the structure itself.
What could be the problem if BeliefPropagation inference takes "too much time"? (I stopped it after a few minutes when VariableElimination finishes under 2 seconds for the same setup)

Now that our network is bigger I experience that the performance is much closer to the non-opensource SMILE library. Even though it's still a bit slower, my main goal would be to handle this use-case faster: https://github.com/pgmpy/pgmpy/discussions/1746#discussioncomment-8984162
From: https://ermongroup.github.io/cs228-notes/inference/jt/

> "Fortunately, it turns out that this problem is also easily avoidable. When computing marginals, VE produces many intermediate factors τ as a side-product of the main computation; these factors turn out to be the same as the ones that we need to answer other marginal queries. By caching them after a first run of VE, we can easily answer new marginal queries at essentially no additional cost.
> The end result of this chapter will be a new technique called the Junction Tree (JT) algorithm, this algorithm will first execute two runs of the VE algorithm to initialize a particular data structure holding a set of pre-computed factors. Once the structure is initialized, it can answer marginal queries in O(1) time.
> We will introduce two variants of this algorithm: belief propagation (BP), and the full junction tree method. BP applies to tree-structured graphs, while the junction-tree method is applicable to general networks."

So here the important part for us is the O(1) time and reusing the calculated factors. In pgmpy docs I see that BeliefPropagation can have a BayesianNetwork model as input. If that is the case then is there an automatic conversion happening to a tree-structured graph? If yes, could you elaborate on that a little? Maybe it's taking so long for our network because the conversion is the problem, not the calculations themselves? (edit: now I see in the code, that if model is not a JunctionTree then to_junction_tree() is called. And now I also see that triangulation seems to be taking a very long time.)
Is my assumption correct that if we could use BeliefPropagation or JunctionTreeAlgorithm then the use-case I linked could be made much faster? So only the first double variable elimination run (+the conversion before) would take long, but then if we just query for different evidences, the actual inference should be really fast? Also if more and more evidence is introduced, would this be the case still, right?

I know that converting a DAG to a tree-structure is hard and elimination order matters a lot. Based on our experience with SMILE library different variable orders when creating the network have huge differences on the created tree graph and hence inference. Does pgmpy have any tools for this? I see to_junction_tree() function, and will experiment with that.

Do you think it would be viable to experiment with all kinds of orderings for a given network that eventually produces a tree graph where maximal clique is ~smaller and hence inference faster?

Sorry for the many questions! :)
Thanks in advance for any answers!

edit: I've experimented a bit with to_junction_tree(). I get this error in one case: 
```
MarkovNetwork.py", line 475, in triangulate
    node_to_delete = min(S_by_C, key=S_by_C.get)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: min() iterable argument is empty
```
What could be the problem? Shouldn't check_model give some error in advance? By the way I think to_junction_tree() is really slow for our network because we are converting noisyMax nodes/variables to regular nodes (before calling to_junction_tree) which adds additional complexity and most likely in a graph like that it's much harder to create a jointree/junctiontree.

**Comment by ankurankan** [comment, marked as answer]
@ferencbartok Hi,

> Is Junction tree algorithm available in pgmpy? I could not find it in the docs, just the structure itself.

The current implementation of Belief Propagation is a variant of the Junction Tree algorithm. It's the exact inference variant, so it tries to construct a junction "tree" and that is the reason why it is inefficient. The approximate versions of the algorithms do not need to construct a tree and hence can have smaller cliques and hence faster.

> Is my assumption correct that if we could use BeliefPropagation or JunctionTreeAlgorithm then the use-case I linked could be made much faster? 

Yes, you should be able to make it faster in your usecase as you can reuse some of the computation that is done in the first step of the algorithm (i.e. calibration). However, as I mentioned above the current bottleneck is because of the constructed junction tree.

> So only the first double variable elimination run (+the conversion before) would take long, but then if we just query for different evidences, the actual inference should be really fast? Also if more and more evidence is introduced, would this be the case still, right?

I will have to double-check this to be sure but I do think it should.

> Do you think it would be viable to experiment with all kinds of orderings for a given network that eventually produces a tree graph where maximal clique is ~smaller and hence inference faster?

I think we would need to switch to the approximate version of the algorithm, so it would need changes in both the tree construction and the calibration method.

**Reply by ferencbartok** [reply]
Thank you for the answers!
So currently the approximate version of junction tree algorithm is not implemented in pgmpy?

Do you have any idea what could have caused this error during triangulation?:
```
          ^^^^^^^^^^^^^^^^^^^^^  
  File "C:\Users\barto\DEV\bayesian-api\venv\Lib\site-packages\pgmpy\models\MarkovNetwork.py", line 525, in to_junction_tree  
    triangulated_graph = self.triangulate()  
                         ^^^^^^^^^^^^^^^^^^  
  File "C:\Users\barto\DEV\bayesian-api\venv\Lib\site-packages\pgmpy\models\MarkovNetwork.py", line 475, in triangulate  
    node_to_delete = min(S_by_C, key=S_by_C.get)  
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^  
ValueError: min() iterable argument is empty
```

Even if calibration takes a day this could still help us (possibly, hopefully) in our use case. But it may take too long, I'll run some tests but I get the above error.

Do you have any suggestions what else we could use instead of junction tree algorithm (for mentioned use case)?

Edit: I'm thinking of changing our algorithm and wondering if you could see any tools in pgmpy that we could use. So the algorithm is basically just setting observable variables (like medical tests) to all of their states one by one and checking their overall combined impact on fault/disease variables (changes in marginal probabilities). For this, in a big network, we have to calculate a ton of inferences with different evidence. Maybe there is an easier way to get a similar answer (to "Which of these nodes observation/evidence would have the biggest impact on those other nodes?") that is already implemented in pgmpy?  (the goal of this is to determine the most important/useful next question towards the user)

**Reply by ankurankan** [reply]
@ferencbartok 
> So currently the approximate version of junction tree algorithm is not implemented in pgmpy?

No. 

> Do you have any idea what could have caused this error during triangulation?:

I am unable to reproduce the error. Can you share some code to reproduce it?

> Even if calibration takes a day this could still help us (possibly, hopefully) in our use case. But it may take too long, I'll run some tests but I get the above error.

Yes, but inefficient tree construction would also slow down the final distribution computation from this calibrated model.

> Do you have any suggestions what else we could use instead of junction tree algorithm (for mentioned use case)?

We had talked about potentially using the sampling-based ApproxInference in the other issue. I don't know if you already gave that a try? I don't know if you have a timeline to get this done but I could also try to add/reimplement our Belief Propagation algorithm to do approx inference. This has already been a long-standing issue, so might be a good time to implement it. 

> Edit: I'm thinking of changing our algorithm and wondering if you could see any tools in pgmpy that we could use. So the algorithm is basically just setting observable variables (like medical tests) to all of their states one by one and checking their overall combined impact on fault/disease variables (changes in marginal probabilities). For this, in a big network, we have to calculate a ton of inferences with different evidence. Maybe there is an easier way to get a similar answer (to "Which of these nodes observation/evidence would have the biggest impact on those other nodes?") that is already implemented in pgmpy? (the goal of this is to determine the most important/useful next question towards the user)

You could compute a joint distribution on both the medical tests and the fault/disease variables (i.e., the evidence and the target variables), and then you can use this joint distribution to compute the conditional distributions that you need. This allows you to perform inference only once but would only work when the total number of evidence + target variables is small enough.

**Reply by ferencbartok** [reply]
> I am unable to reproduce the error. Can you share some code to reproduce it?

It's nothing special, just creating the model and then using BeliefPropagation(model)

```
model = BayesianNetwork()
many model.add_node(node.id)
many model.add_edge(parent_id, node.id)
model.add_cpds(*node_id_to_CPD.values())
then
bp = BeliefPropagation(model)
```
I think it is because of our network. But if our network is wrong, check_model should throw an error I think. So maybe some edge case is not covered in to_junction_tree()?

> Yes, but inefficient tree construction would also slow down the final distribution computation from this calibrated model.

Why is that? I thought that once we have an almost optimal tree then the marginal queries would be fast. Or maybe I misunderstand what you mean. What do you mean by the "final distribution computation"?

> We had talked about potentially using the sampling-based ApproxInference in the other issue. I don't know if you already gave that a try? I don't know if you have a timeline to get this done but I could also try to add/reimplement our Belief Propagation algorithm to do approx inference. This has already been a long-standing issue, so might be a good time to implement it.

Yes, I've tried it back then but it was slow or not precise if I remember correctly. Now my plan is to generate a huge sample and run adding more and more evidence and see it's behaviour.
Well, the timeline is hard to describe as we are searching for different solutions in all domains (better CPUs, different kinds of parallelism, different node orderings, modifying networks, creating new algorithm and so on). Of course, we would be happy if this feature was under development, but don't know if in the end, it would solve our problems. But thanks if you start it and if not also thanks for the answers! :)

> You could compute a joint distribution on both the medical tests and the fault/disease variables (i.e., the evidence and the target variables), and then you can use this joint distribution to compute the conditional distributions that you need. This allows you to perform inference only once but would only work when the total number of evidence + target variables is small enough.

I understand, I will try this although I have a suspicion we already have too many variables. Back when I first tested pgmpy I ran into memory problems when calculating full joint distribution. Thanks for the tip!

**Reply by ankurankan** [reply]
> Why is that? I thought that once we have an almost optimal tree then the marginal queries would be fast. Or maybe I misunderstand what you mean. What do you mean by the "final distribution computation"?

If we have an optimal tree, then it would be fast. I meant for the case when the tree isn't optimal (i.e., the cliques are quite large), in that case computing the queried distribution can still be slow because the factors of these large cliques would be huge.

**Reply by ferencbartok** [reply]
I understand, thanks.
I'm wondering what techniques we could use from pgmpy to possibly make the graph/network simpler without losing too much information (or none). What I mean is possibly deleting edges and/or nodes. Would you have some suggestions regarding this possibility in order to speed up inference? (possibly using d-separation utilities or structure learning maybe?)

**Reply by ankurankan** [reply]
We already do d-separation based network pruning based on query and evidence variables: https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/inference/base.py#L120 . Another idea could be to remove edges from the network whose edge strength is below some threshold. You can compute a canonical correlation based edge weight like:
```python
In [7]: from pgmpy.utils import get_example_model

In [8]: model = get_example_model('alarm')

In [9]: df = model.simulate(1000)
Generating for node: BP:   0%|                                                                                                                                         | 0/37 [00:00<?, ?it/s]WARNING:pgmpy:Probability values don't exactly sum to 1. Differ by: -2.220446049250313e-16. Adjusting values.
Generating for node: BP: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 37/37 [00:00<00:00, 714.30it/s]

In [10]: from pgmpy.estimators.CITests import pillai_trace
In [11]: for u, v in model.edges():
    ...:     u_parent = set(model.get_parents(u))
    ...:     v_parent = set(model.get_parents(v))
    ...:     cond_vars = u_parent.union(v_parent)
    ...:     print(f"Edge {u} -> {v} weight: {pillai_trace(X=u, Y=v, Z=cond_vars, data=df, boolean=False)[0]}")
    ...: 
Edge HYPOVOLEMIA -> LVEDVOLUME weight: 2.8810900939807843e-06
Edge HYPOVOLEMIA -> STROKEVOLUME weight: 6.5546008480529146e-06
Edge LVEDVOLUME -> CVP weight: 2.9986980925301876e-06
Edge LVEDVOLUME -> PCWP weight: 1.856005839558033e-05
Edge LVFAILURE -> HISTORY weight: 4.982297539629375e-09
Edge LVFAILURE -> LVEDVOLUME weight: 6.199002108034396e-06
```

**Reply by ferencbartok** [reply]
Thank you, I'll experiment with this as well.
Another question regarding the possible implementation of the approximate variant of ~junction tree algorithm: do you mean implementing loopy belief propagation or some other algorithm?

**Reply by ankurankan** [reply]
Yes, I was thinking something along the lines of loopy BP. However, I am not very updated on the literature if there are any better variants to it.

**Reply by ferencbartok** [reply]
When running the pillai_trace I get the following error:
```
Error
Traceback (most recent call last):
    print(f"Edge {u} -> {v} weight: {ci_pillai(X=u, Y=v, Z=cond_vars, data=df, boolean=False)[0]}")
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\venv\Lib\site-packages\pgmpy\estimators\CITests.py", line 793, in ci_pillai
    res_x_c, res_y_c = cca.fit_transform(res_x, res_y)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\venv\Lib\site-packages\sklearn\utils\_set_output.py", line 316, in wrapped
    data_to_wrap = f(self, X, *args, **kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\venv\Lib\site-packages\sklearn\cross_decomposition\_pls.py", line 552, in fit_transform
    return self.fit(X, y).transform(X, y)
           ^^^^^^^^^^^^^^
  File "\venv\Lib\site-packages\sklearn\base.py", line 1473, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\venv\Lib\site-packages\sklearn\cross_decomposition\_pls.py", line 393, in fit
    pinv2(np.dot(self.x_loadings_.T, self.x_weights_), check_finite=False),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\venv\Lib\site-packages\scipy\_lib\deprecation.py", line 213, in inner_f
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "\venv\Lib\site-packages\scipy\linalg\_basic.py", line 1423, in pinv
    u, s, vh = _decomp_svd.svd(a, full_matrices=False, check_finite=False)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\venv\Lib\site-packages\scipy\linalg\_decomp_svd.py", line 147, in svd
    raise ValueError('illegal value in %dth argument of internal gesdd'
ValueError: illegal value in 4th argument of internal gesdd
```

Do you have any idea what could cause this?

**Reply by ankurankan** [reply]
@ferencbartok Very hard to say without looking at the data. Are all the variables continuous? Is either X or Y in Z ?

**Reply by ferencbartok** [reply]
All variables are discrete, I guess then it is a problem as it seems pillai test is for continuous variables?

**Reply by ankurankan** [reply]
No, pillai test works for any variable type. Is there any chance that one of your variables has only 1 state, or there are duplicated columns?
